flang/lib/Lower/PFTBuilder.cpp

```diff
@@ -1291,29 +1322,20 @@ struct SymbolDependenceDepth {
       // TODO: add declaration?
       return 0;
     }
-    if (sym.has<semantics::UseDetails>() ||
-        sym.has<semantics::HostAssocDetails>() ||
-        sym.has<semantics::NamelistDetails>() ||
-        sym.has<semantics::ModuleDetails>() ||
-        sym.has<semantics::MiscDetails>()) {
-      // FIXME: do we want to do anything with any of these?  Other syms?
+    auto ultimate = sym.GetUltimate();
+    if (!ultimate.has<semantics::ObjectEntityDetails>() &&
+        !ultimate.has<semantics::ProcEntityDetails>())
       return 0;
-    }
 
     if (sym.has<semantics::DerivedTypeDetails>())
       llvm_unreachable("not yet implemented - derived type analysis");
 
     // Symbol must be something lowering will have to allocate.
     bool global = semantics::IsSaved(sym);
-    int depth = 0;
+    int depth = 1; // 0 is reserved for aggregates store
```

Comment:
That's a fair point, my motivation to do so was to avoid having to declare unused symbols.
With modules, the "stores" variable may now contains unused aggregate store (`analyseAliases` builds intervals for all aliases in a modules, not only for the one used. It would make `analyseAliases` more complex to determine if stores are needed in the scope using the module). However, that's probably not a big deal to have unsued store variables being declared.

Illustration:

```
module m
 real :: x, y
 integer :: i
 equivalence (i, y)
end module

! In another file
subroutine foo
  use m, only: x
  print *, x
end subroutine
```

In the above example, the point of delaying putting the stores in the variables was to avoid declaring the storage for `equivalence (i, y)` when lowering foo. But I have to admit that apart that it is a bit weird, it's not a big deal.

But I think your point that it yields worse handling for normal case and that equivalence are rare is fair. I will revert that.

Reply:
Yes, the scheme did/does rely on dead allocations getting removed. I thought DCE did a decent job of that, but I could be wrong.